sgkit/distance/metrics.py

```diff
@@ -172,3 +173,46 @@ def correlation_reduce(v: ArrayLike, out: ArrayLike) -> None:  # pragma: no cove
     if denom > 0:
         value = 1 - (num / denom)
     out[0] = value
+
+
+@cuda.jit(device=True)
+def _euclidean_distance(a, b):
+    square_sum = 0.0
+    for i in range(a.shape[0]):
+        if a[i] >= 0 and b[i] >= 0:
+            square_sum += (a[i] - b[i]) ** 2
+    return square_sum
+
+
+@cuda.jit
+def euclidean_map_kernel(x, y, out) -> None:
+    i1, i2 = cuda.grid(2)
+    if i1 >= x.shape[0] or i2 >= y.shape[0]:
+        # Quit if (x, y) is outside of valid output array boundary
+        return
+    out[i1][i2] = _euclidean_distance(x[i1], y[i2])
+
+
+def euclidean_map_gpu(f, g):
+    # move input data to the device
+    d_a = cuda.to_device(f)
+    d_b = cuda.to_device(g)
+    # create output data on the device
+    out = np.zeros((f.shape[0], g.shape[0]))
+    #     print(f"out_shape: {out.shape}")
+    d_out = cuda.to_device(out)
```

Comment:
I started with `cupy` initially, but there was something I don't recall that was not implemented in it, so it didn't work. But I think the code has changed a lot since then, so I guess that could work, thanks, I'll try that.